File: src/main/java/de/tomjuri/armageddon/util/RotationUtil.java
package de.tomjuri.armageddon.util;

import de.tomjuri.armageddon.Armageddon;
import lombok.AllArgsConstructor;
import lombok.Getter;
import net.minecraft.util.MathHelper;

public class RotationUtil {
    private static Rotation startRotation = new Rotation(0f, 0f);
    private static Rotation endRotation = new Rotation(0f, 0f);
    private static long startTime = 0L;
    private static long endTime = 0L;
    private static boolean done = true;

    public static void ease(Rotation rotation, long durationMillis) {
        if (!done) return;
        done = false;
        startRotation = new Rotation(Armageddon.INSTANCE.getPlayer().rotationYaw, Armageddon.INSTANCE.getPlayer().rotationPitch);
        Rotation neededChange = getNeededChange(startRotation, rotation);
        endRotation = new Rotation(startRotation.getYaw() + neededChange.getYaw(), startRotation.getPitch() + neededChange.getPitch());
        startTime = System.currentTimeMillis();
        endTime = startTime + durationMillis;
    }

    public static void easeCertain(Rotation rotation, long durationMillis) {
        if (!done) return;
        done = false;
        Rotation currentRotation = new Rotation(Armageddon.INSTANCE.getPlayer().rotationYaw, Armageddon.INSTANCE.getPlayer().rotationPitch);
        float endRotationYaw = currentRotation.getYaw() + rotation.getYaw();
        startRotation = currentRotation;
        endRotation = new Rotation(endRotationYaw, rotation.getPitch());
        startTime = System.currentTimeMillis();
        endTime = startTime + durationMillis;
    }

    public static void onRenderWorldLast() {
        if (done) return;
        if (System.currentTimeMillis() <= endTime) {
            Armageddon.INSTANCE.getPlayer().rotationYaw = interpolate(startRotation.getYaw(), endRotation.getYaw());
            Armageddon.INSTANCE.getPlayer().rotationPitch = interpolate(startRotation.getPitch(), endRotation.getPitch());
            return;
        }
        Armageddon.INSTANCE.getPlayer().rotationYaw = endRotation.getYaw();
        Armageddon.INSTANCE.getPlayer().rotationPitch = endRotation.getPitch();
        done = true;
    }

    private static float interpolate(float start, float end) {
        float spentMillis = (float) (System.currentTimeMillis() - startTime);
        float relativeProgress = spentMillis / (endTime - startTime);
        return (end - start) * easeOutCubic(relativeProgress) + start;
    }

    private static float easeOutCubic(float number) {
        return (float) (1.0 - Math.pow(1.0 - number, 3.0));
    }

    private static Rotation getNeededChange(Rotation startRot, Rotation endRot) {
        float yawChange = MathHelper.wrapAngleTo180_float(endRot.getYaw()) - MathHelper.wrapAngleTo180_float(startRot.getYaw());
        if (yawChange <= -180.0f) yawChange += 360.0f;
        else if (yawChange > 180.0f) yawChange += -360.0f;
        return new Rotation(yawChange, endRot.getPitch() - startRot.getPitch());
    }

    @Getter
    @AllArgsConstructor
    public static class Rotation {
        private float yaw;
        private float pitch;
    }
}

File: src/main/kotlin/de/tomjuri/armageddon/util/RotationUtil.kt
package de.tomjuri.armageddon.util

import net.minecraft.util.MathHelper
import kotlin.math.pow


object RotationUtil {
    private var startRotation = Rotation(0f, 0f)
    private var endRotation = Rotation(0f, 0f)
    private var startTime = 0L
    private var endTime = 0L
    private var done = true

    fun ease(rotation: Rotation, durationMillis: Long) {
        if (!done) return
        done = false
        startRotation = Rotation(player.rotationYaw, player.rotationPitch)
        val neededChange = getNeededChange(startRotation, rotation)
        endRotation = Rotation(startRotation.yaw + neededChange.yaw, startRotation.pitch + neededChange.pitch)
        startTime = System.currentTimeMillis()
        endTime = startTime + durationMillis
    }

    fun easeDirection(rotation: Rotation, durationMillis: Long, direction: Direction) {
        if (!done) return
        done = false
        val currentRotation = Rotation(player.rotationYaw, player.rotationPitch)
        val endRotationYaw: Float = if(direction == Direction.LEFT) currentRotation.yaw - rotation.yaw else currentRotation.yaw + rotation.yaw
        startRotation = currentRotation
        endRotation = Rotation(endRotationYaw, rotation.pitch)
        startTime = System.currentTimeMillis()
        endTime = startTime + durationMillis
    }

    fun onRenderWorldLast() {
        if (done) return
        if (System.currentTimeMillis() <= endTime) {
            player.rotationYaw = interpolate(startRotation.yaw, endRotation.yaw)
            player.rotationPitch = interpolate(startRotation.pitch, endRotation.pitch)
            return
        }
        player.rotationYaw = endRotation.yaw
        player.rotationPitch = endRotation.pitch
        done = true
    }

    private fun interpolate(start: Float, end: Float): Float {
        val spentMillis = (System.currentTimeMillis() - startTime).toFloat()
        val relativeProgress = spentMillis / (endTime - startTime)
        return (end - start) * easeOutCubic(relativeProgress) + start
    }

    private fun easeOutCubic(number: Float) = (1.0 - (1.0 - number).pow(3.0)).toFloat()

    private fun getNeededChange(startRot: Rotation, endRot: Rotation): Rotation {
        var yawChange = MathHelper.wrapAngleTo180_float(endRot.yaw) - MathHelper.wrapAngleTo180_float(startRot.yaw)
        if (yawChange <= -180.0f) yawChange += 360.0f else if (yawChange > 180.0f) yawChange += -360.0f
        return Rotation(yawChange, endRot.pitch - startRot.pitch)
    }

    data class Rotation(val yaw: Float, val pitch: Float)
    enum class Direction {
        LEFT,
        RIGHT
    }
}